core/lib/multivm/src/versions/vm_fast/vm.rs

```diff
@@ -596,18 +644,9 @@ impl<S: ReadStorage, Tr: Tracer + Default> Vm<S, Tr> {
         let start = self.inner.world_diff().snapshot();
         let gas_before = self.gas_remaining();
 
-        let mut full_tracer = (
-            (mem::take(tracer), CircuitsTracer::default()),
-            EvmDeployTracer::new(self.world.dynamic_bytecodes.clone()),
-        );
-        let result = self.run(
-            execution_mode,
-            &mut full_tracer,
-            track_refunds,
-            pubdata_builder,
-        );
-        let ((external_tracer, circuits_tracer), _) = full_tracer;
-        *tracer = external_tracer; // place the tracer back
+        tracer.insert_dynamic_bytecodes_handle(self.world.dynamic_bytecodes.clone());
```

Comment:
I didn't really like the `'static` in the interface and the fact that you pass in a reference but the contents of it then get moved. I think I'd prefer a solution where the reference is stored over the move.

You are right that this could cause tracers to be incorrectly reused which is something I did not think about.

IIRC this was also done so the validation result can be taken out of the tracer without extra synchronization primitives.